cpp/src/io/parquet/arrow_schema_writer.cpp

```diff
@@ -0,0 +1,379 @@
+    field_offset = flatbuf::CreateInt(fbb, 64, std::numeric_limits<T>::is_signed).Union();
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_same_v<T, float>, void> operator()()
+  {
+    type_type    = flatbuf::Type_FloatingPoint;
+    field_offset = flatbuf::CreateFloatingPoint(fbb, flatbuf::Precision::Precision_SINGLE).Union();
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_same_v<T, double>, void> operator()()
+  {
+    type_type    = flatbuf::Type_FloatingPoint;
+    field_offset = flatbuf::CreateFloatingPoint(fbb, flatbuf::Precision::Precision_DOUBLE).Union();
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_same_v<T, cudf::string_view>, void> operator()()
+  {
+    type_type    = flatbuf::Type_Utf8View;
+    field_offset = flatbuf::CreateUtf8View(fbb).Union();
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_same_v<T, cudf::timestamp_D>, void> operator()()
+  {
+    type_type = flatbuf::Type_Date;
+    // Use one of the strings: "UTC", "Etc/UTC" or "+00:00" to indicate a native UTC timestamp
+    field_offset = flatbuf::CreateDate(fbb, flatbuf::DateUnit_DAY).Union();
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_same_v<T, cudf::timestamp_s>, void> operator()()
+  {
+    type_type = flatbuf::Type_Timestamp;
+    // Use one of the strings: "UTC", "Etc/UTC" or "+00:00" to indicate a native UTC timestamp
+    field_offset = flatbuf::CreateTimestamp(
+                     fbb, flatbuf::TimeUnit_SECOND, (utc_timestamps) ? fbb.CreateString("UTC") : 0)
+                     .Union();
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_same_v<T, cudf::timestamp_ms>, void> operator()()
+  {
+    type_type = flatbuf::Type_Timestamp;
+    // Use one of the strings: "UTC", "Etc/UTC" or "+00:00" to indicate a native UTC timestamp
+    field_offset =
+      flatbuf::CreateTimestamp(
+        fbb, flatbuf::TimeUnit_MILLISECOND, (utc_timestamps) ? fbb.CreateString("UTC") : 0)
+        .Union();
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_same_v<T, cudf::timestamp_us>, void> operator()()
+  {
+    type_type = flatbuf::Type_Timestamp;
+    // Use one of the strings: "UTC", "Etc/UTC" or "+00:00" to indicate a native UTC timestamp
+    field_offset =
+      flatbuf::CreateTimestamp(
+        fbb, flatbuf::TimeUnit_MICROSECOND, (utc_timestamps) ? fbb.CreateString("UTC") : 0)
+        .Union();
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_same_v<T, cudf::timestamp_ns>, void> operator()()
+  {
+    type_type = flatbuf::Type_Timestamp;
+    // Use one of the strings: "UTC", "Etc/UTC" or "+00:00" to indicate a native UTC timestamp
+    field_offset =
+      flatbuf::CreateTimestamp(
+        fbb, flatbuf::TimeUnit_NANOSECOND, (utc_timestamps) ? fbb.CreateString("UTC") : 0)
+        .Union();
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_same_v<T, cudf::duration_D>, void> operator()()
+  {
+    type_type    = flatbuf::Type_Time;
```

Comment:
duration_D is written as `TimeType` as `duration_D` is not a valid arrow type. This also allows for easy and faithful roundtripping with cudf.

Reply:
please leave a comment